### PR TITLE
[Integration][AzureDevops] Added Support for Glob Searching in ADO File Kind

### PR DESCRIPTION
**Pull request by emekanwaoma**
### **User description**
# Description

What - Enhanced the file kind in AzureDevopsClient to support glob pattern matching and efficient file retrieval from Azure DevOps repositories.

Why

- The existing file generation logic only handled literal file paths and lacked flexibility in pattern-based discovery. This enhancement allows users to: 
   - Retrieve files using glob-style patterns (e.g., **/*.yaml)
   - Efficiently fetch and filter files across repositories
   - Support deeper content scanning scenarios with better concurrency and control

How
- Introduced support for glob pattern parsing and descriptor extraction
- Used itemsbatch API to fetch files based on PathDescriptor objects
- Grouped file patterns by base paths to minimize API requests
- Filtered matching files post-fetch using glob rules

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Added glob pattern support for file path matching

- Enhanced file retrieval with itemsbatch API optimization

- Improved file filtering and processing capabilities

- Updated webhook processing for glob pattern matching

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["File Paths"] --> B["Path Separation"]
  B --> C["Literal Paths"]
  B --> D["Glob Patterns"]
  C --> E["Direct API Fetch"]
  D --> F["Pattern Descriptor"]
  F --> G["Grouped API Requests"]
  E --> H["File Processing"]
  G --> I["Glob Filtering"]
  I --> H
  H --> J["Downloaded Files"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>azure_devops_client.py</strong><dd><code>Enhanced file retrieval with glob support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/azure_devops/client/azure_devops_client.py

<ul><li>Refactored <code>_get_repository_files</code> to support both literal and glob <br>patterns<br> <li> Added <code>_get_files_by_explicit_paths</code> and <code>_get_files_by_descriptors</code> <br>methods<br> <li> Integrated glob pattern processing with descriptor-based file fetching<br> <li> Enhanced error handling for itemsbatch API requests</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-6d76e1ab85e24b8f643174f9a502ef2dd04e0dc019a5ebd6b1c8e7367624da4f">+85/-41</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>file_processing.py</strong><dd><code>Core glob pattern processing utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/azure_devops/client/file_processing.py

<ul><li>Added <code>PathDescriptor</code> class and <code>RecursionLevel</code> enum for structured path <br>handling<br> <li> Implemented glob pattern detection with <code>is_glob_pattern</code> function<br> <li> Added descriptor extraction and grouping utilities<br> <li> Created file filtering functions for glob pattern matching</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-e65d9c5e39277b1cbb3b92067ac36143c25e2a3cd00466f53163c3353ce4409b">+113/-1</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>file_webhook_processor.py</strong><dd><code>Updated webhook processing for glob patterns</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/azure_devops/webhooks/webhook_processors/file_webhook_processor.py

<ul><li>Replaced <code>fnmatch</code> with <code>matches_glob_pattern</code> for consistent glob <br>handling<br> <li> Updated file path processing to strip leading slashes<br> <li> Enhanced webhook file change detection with glob pattern support</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-8b53fa098bf55e9f1357fb393ac94787f376d6bc6c86b80f4c96185213c59a21">+11/-5</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>integration.py</strong><dd><code>Enabled glob patterns in configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/integration.py

<ul><li>Removed path validation that prohibited glob patterns<br> <li> Updated <code>FilePattern</code> documentation to include glob pattern examples<br> <li> Simplified path handling by removing restrictive validators</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-91fe285633f234d3bb646a0c394e2303b0100c2cf3dbc712c9c1345128b72b50">+11/-46</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_azure_devops_client.py</strong><dd><code>Comprehensive glob pattern testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/tests/azure_devops/client/test_azure_devops_client.py

<ul><li>Added comprehensive tests for glob pattern file generation<br> <li> Created test cases for mixed literal and glob patterns<br> <li> Added tests for different recursion levels and edge cases<br> <li> Implemented mock scenarios for various file matching scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-bd6aae4af7c82e93c446c23fc8198110a7e29f73e2fe861ce9940c7a079bacf5">+551/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_file_processing.py</strong><dd><code>File processing utility tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/tests/azure_devops/client/test_file_processing.py

<ul><li>Added extensive test coverage for new glob processing functions<br> <li> Created test classes for pattern detection, descriptor extraction, and <br>file filtering<br> <li> Implemented tests for edge cases and complex glob patterns</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-8a083e6bae63f311e841eab648334cfc504814a9d90e5a3d935382cc25ae81b0">+345/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Version 0.2.35 changelog update</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/CHANGELOG.md

<ul><li>Added changelog entry for version 0.2.35<br> <li> Documented glob pattern support and itemsbatch API optimization <br>improvements</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-9b4f288c83afcf2c6cdce8a77deca105f812a5b9af7b7b6125bce0786dbfd6a6">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Added wcmatch dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/azure-devops/pyproject.toml

- Added `wcmatch` dependency for advanced glob pattern matching

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1954/files#diff-067c63eae2423648d5f4f587433376d049963f0e9e2a57eb485817e5deb32f5a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/azure-devops/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+## 0.2.35 (2025-08-05)
+
+
+## Improvements
+
+- Added glob pattern support in AzureDevopsClient
```
Comment (matan84) [inline integrations/azure-devops/CHANGELOG.md]: Add an explanation of what needs to be added to the mapping in order to use it
